### PR TITLE
294755 - Build search results query based on specific date ranges

### PR DESCRIPTION
**Pull request by FujaleAhmedShaik**
### What one thing this PR does?

Search results can now be retrieved by filtering by date range. This is the first step of guided search filtering.

### Task details

- [294755 - Build search results based on specific date ranges.](https://dev.azure.com/defragovuk/DEFRA-NCEA/_workitems/edit/294755)

### Other notes

- Now we can retrieve the records by date range. Validation of dates is available in another feature [PR](https://github.com/DEFRA/ncea-frontend/pull/11).
- As part of this ticket, we have also made many improvements to the code.
- We have refactored the search API query to build freely based on the search term, date range, and geoshape coordinate. The depth provided will also be taken into consideration.
- Now we are searching in all available fields to find the given search term, but in the future we may have to restrict it to specific fields. However, this is already covered as part of this ticket.
- Also, we were able to achieve unit test coverage of almost 95%.

### How do these changes look like?

**Quick Search**
| Quick search |
| ------------------------------ |

**Quick Search results**
| Quick search results |
| ------------------------------ |

**Guided Search date filter**
| Guided Search date filter |
| ------------------------------ |

**Guided Search date filter results**
| Guided Search date filter results |
| ------------------------------ |

**Test Coverage**
| Test Coverage |
| ------------------------------ |

### Dev-tested in

1. Local environment

- [Quick search](http://localhost:3000/)
- [Guided search date filter](http://localhost:3000/date-search)
- [Search results](http://localhost:3000/searchh)

2. Dev environment

- [Quick search](http://localhost:3000/)
- [Guided search date filter](http://localhost:3000/date-search)
- [Search results](http://localhost:3000/searchh)

### CODE REVIEW

**Comment by karthigayayiniram** [comment]
changed as Draft PR to stop accidental merges as it has to sequentially merged as per @FujaleAhmedShaik comments

**Review by ganesh-rajamahanthy — APPROVED** [review]

**Review by ganesh-rajamahanthy — DISMISSED** [review]

File: src/infrastructure/plugins/shared_data.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { ISharedData } from '../../interfaces/sharedData.interface';
+import { Server } from '@hapi/hapi';
+
+module.exports = {
+  name: 'SharedData',
```
Comment (karthigayayiniram) [inline src/infrastructure/plugins/shared_data.ts:5]: i personally think this plugin is not going workout when you the scale the application horizontally. i prefer not to use any cache plugin as we are using elastic search. if you still want to have one, please think about using redis or memcache

**Comment by karthigayayiniram** [comment]
Renaming the files that have been merged earlier increases the number of files to be reviewed, obscuring the original intended files. It should instead be added as a PR comment to the respective PR

File: src/controllers/web/SampleController.ts [hunk 2]
```diff
@@ -1,12 +1,8 @@
...
-// import { getSamplePosts } from '../../services/handlers/samplePostApi';
 import { Request, ResponseObject, ResponseToolkit } from '@hapi/hapi';
 
 const SampleController = {
   renderSampleHandler: async (request: Request, response: ResponseToolkit): Promise<ResponseObject> => {
-    // const posts = await getSamplePosts();
-
     return response.view('sample', {
       title: 'Hapi ' + request.server.version,
       message: 'Hello Nunjucks!',
```
[the remark this answers is not in the document]
Comment (FujaleAhmedShaik) [reply src/controllers/web/SampleController.ts]: Removed the sample files

File: src/routes/web/searchResults.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+import { SearchResultsController } from '../../controllers/web/SearchResultsController';
+import { webRoutePaths } from '../../utils/constants';
+
+const searchResultsRoutes = [
+  {
+    method: 'GET',
+    path: webRoutePaths.results,
+    handler: SearchResultsController.renderSearchResultsHandler,
```
[the remark this answers is not in the document]
Comment (FujaleAhmedShaik) [reply src/routes/web/searchResults.ts:10]: For now I'll keep this  as is. Once your back we can discuss on this.

File: public/scripts/fetchResults.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      storeStorageData(sessionData);
+    }
+  });
+};
+
+document.addEventListener('DOMContentLoaded', () => {
+  const fetchResults = document.querySelector('[data-fetch-results]');
+  if (fetchResults) {
+    const action = fetchResults.getAttribute('data-action');
+    getSearchResults(action);
```
Comment (karthigayayiniram) [inline public/scripts/fetchResults.js:78]: you should also be checking if there are no form error.
image

**Review by karthigayayiniram — COMMENTED** [review]
Not sure my local has problem. i can't see search results and getting 500 internal server error. also it would be great to see a loader till the ajax call fetches the partial html and renders it on the ui.
ajax-fails
image

File: __tests__/routes/web/searchResults.spec.ts [hunk 5]
```diff
@@ -0,0 +1,549 @@
+/**
+ * @jest-environment jsdom
+ */
+
+'use strict';
+
+import { Server, ServerInjectResponse } from '@hapi/hapi';
+
+import { initializeServer } from '../../../src/infrastructure/server';
+import supertest from 'supertest';
```
Comment (karthigayayiniram) [inline __tests__/routes/web/searchResults.spec.ts:10]: why are you introducing a new library now? Has this been approved by somebody?

Comment (FujaleAhmedShaik) [reply __tests__/routes/web/searchResults.spec.ts:10]: About the supertest package. Using server.inject takes a lot of time and causes a request timeout error while testing the POST request. Also, the assertions placed for POST request pass even if they are not correct. So I included this to be used in the test cases.

File: src/utils/generateDateString.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+const generateDateString = (dateObject: IDateObject): string => {
+  const month = !isEmpty(dateObject.month) && !isNaN(dateObject.month!) ? dateObject.month! : 1;
+  const day = !isEmpty(dateObject.day) && !isNaN(dateObject.day!) ? dateObject.day! : 1;
+
+  const date = new Date(Date.UTC(dateObject.year, month - 1, day));
+
+  const dateString = date.toISOString().slice(0, 10);
+
+  return dateString;
```
[the remark this answers is not in the document]
Comment (FujaleAhmedShaik) [reply src/utils/generateDateString.ts:22]: Yes, now we have implemented that. Included a test case for this as well.

File: src/utils/generateDateString.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { IDateObject } from '../interfaces/searchPayload.interface';
+import { isEmpty } from './isEmpty';
+
+const generateDateString = (dateObject: IDateObject, isToDate = false): string => {
+  const currentDate = new Date();
+  const month = !isEmpty(dateObject.month) && !isNaN(dateObject.month!) ? dateObject.month! : 1;
+  let day: number;
+  if (isToDate && (isEmpty(dateObject.day) || isNaN(dateObject.day!))) {
+    if (month === currentDate.getMonth() + 1) {
```
Comment (karthigayayiniram) [inline src/utils/generateDateString.ts:9]: you should also be checking for the year is present year.
eg: 
day - 19
month - empty
year - 2024
then the end date should be 16-02-2024
but looks like this code returns current day for any feb month irrespective of the year.

Comment (FujaleAhmedShaik) [reply src/utils/generateDateString.ts]: Thanks for pointing this out. Now it has been addressed.

**Review by karthigayayiniram — APPROVED** [review]

**Review by ganesh-rajamahanthy — APPROVED** [review]

**Pull request merged**